File: rust/fuzz/fuzz_targets/format_source.rs
#![no_main]

use libfuzzer_sys::fuzz_target;

fuzz_target!(|data: &[u8]| {
    let Ok(source) = std::str::from_utf8(data) else {
        return;
    };

    let Ok(formatted) = scon::format_source(source, scon::FormatOptions::default()) else {
        return;
    };

    scon::parse_source(&formatted, scon::ParseOptions::default())
        .expect("formatted SCON should parse");

    if let (Ok(original), Ok(round_trip)) = (scon::parse_str(source), scon::parse_str(&formatted)) {
        assert_eq!(original, round_trip, "formatting changed resolved value");
    }
});

File: swift/Fuzz/format_source.swift
import Foundation

@_cdecl("LLVMFuzzerTestOneInput")
public func LLVMFuzzerTestOneInput(_ data: UnsafePointer<UInt8>, _ size: Int) -> Int32 {
    let bytes = UnsafeBufferPointer(start: data, count: size)
    guard let source = String(bytes: bytes, encoding: .utf8) else { return 0 }

    let formatted: String
    do {
        formatted = try Scon.formatSource(source)
    } catch {
        return 0
    }

    if Scon.analyzeSource(formatted).parsed == nil {
        fatalError("formatted source does not parse")
    }

    do {
        let original = try Scon.parseString(source)
        let roundTrip = try Scon.parseString(formatted)
        if try Scon.formatValue(original) != Scon.formatValue(roundTrip) {
            fatalError("formatted source changed resolved value")
        }
    } catch {}

    return 0
}
